### DISCUSSION TITLE
Logging installed binary version and constraints mismatch

### DISCUSSION THREAD

**Discussion by atomrc** [renovatebot/renovate#24920] [Suggest an Idea]
**What would you like Renovate to be able to do?**

Lately I was trying to debug binary version constraints that seemed to be never applied to our self hosted version of renovate. (cf. https://github.com/renovatebot/renovate/discussions/10376)

I wanted to use `npm > 7` but, since we are using the `renovate/renovate` docker container without the `--binary-source=docker` and no `docker.sock` available, there was no way for renovate to be able to install a different `npm` version as the one installed on the system. 

**Did you already have any implementation ideas?**

To debug this, I needed to dig into the code, the fact that there is a mismatch between the constraints and the actual version that is going to be used to install the packages appears **nowhere in the logs**. 

So I would like to suggest adding an entry in the log (probably a warning) that indicates a version of a binary is needed but cannot be installed (maybe suggesting using the `--binary-source + docker.sock` options to fix this. 

In the case of `npm`, the log could probably live around  those lines https://github.com/renovatebot/renovate/blob/9ca604b3453ca55795bd10d3fcad4475c3c8c7cb/lib/util/exec/index.ts#L124-L139
I do not know where it could live in the case of a generic log that would print a mismatch for any binary... 🤔 

WDYT?

**Comment by rarkins** [comment]
Renovate currently cares about constraints when `binarySource=docker`, but mostly ignores them if `binarySource=global`. It's used:

https://github.com/renovatebot/renovate/blob/d10b558beb9e3a6bea96cf95043355cd92faadc9/lib/manager/npm/post-update/npm.ts#L29

https://github.com/renovatebot/renovate/blob/d10b558beb9e3a6bea96cf95043355cd92faadc9/lib/manager/npm/post-update/npm.ts#L34

https://github.com/renovatebot/renovate/blob/d10b558beb9e3a6bea96cf95043355cd92faadc9/lib/manager/npm/post-update/npm.ts#L44

https://github.com/renovatebot/renovate/blob/d10b558beb9e3a6bea96cf95043355cd92faadc9/lib/manager/npm/post-update/npm.ts#L66-L71

This file doesn't know/care about the value of `binarySource`, so the options are:

1. Make it care about `binarySource`, and if it's `global` while `npmConstraint` is defined, then do a check of `npm --version` and warn if it's not satisfying
2. Refactor the exec interface so that it is passed the constraint not just for Docker but also for global and it does the validation. The problem is that it's passed a *full command* and doesn't know what the "version" command is or even that `npm` is the command.

**Comment by rarkins** [comment]
This might be easier once we support binarySource=install for all tools as we can pass that through to `exec`
